File: Priority CPU Scheduling.cpp
#include<bits/stdc++.h>
using namespace std;
#define v vector

void print(v<v<int>> &arr){
    cout<<"Sorted matrix:\n";
    for(int i=0;i<arr.size();i++){
        for(int j=0;j<arr[0].size();j++){
            cout<<arr[i][j]<<" ";
        }
        cout<<"\n";
    }
    cout<<"\n";
}

bool sortHelper(const v<int>&v1,const v<int>&v2){
    return v1[2]>v2[2];
}

void findWaiting(v<v<int>> &arr){
    arr[0][3] = 0;
    for(int i=1;i<arr.size();i++){
        arr[i][3] = arr[i-1][1] + arr[i-1][3];
    }
}

void turnAroundTime(v<v<int>> &arr){
    for(int i=0;i<arr.size();i++){
        arr[i][4] = arr[i][1] + arr[i][3];
    }
}

void priorityScheduling(v<v<int>> &arr){
    sort(begin(arr),end(arr),sortHelper);

    findWaiting(arr);
    turnAroundTime(arr);
    int tot_wat = 0,tot_tat=0;
    cout<<"Process || Bust || Priority || Waiting || Turn Around \n";
    for(int i=0;i<arr.size();i++){
        tot_wat+=arr[i][3];
        tot_tat+=arr[i][4];
        cout<<"   "<<arr[i][0]<<"\t   "<<arr[i][1]<<"\t     "<<arr[i][2]<<"\t\t"<<arr[i][3]<<"\t\t"<<arr[i][4]<<"\n";
    }
    cout<<"Average Waiting Time:->"<<(float)tot_wat/arr.size()<<"\n";
    cout<<"Average Turn Around Time:->"<<(float)tot_tat/arr.size()<<"\n";
    //print(arr);
}


int main(){
    cout<<"Enter Total Number of Process:\n";
    int n;cin>>n;
    v<v<int>> arr;
    cout<<"\"Enter All Details\"\n";
    for(int i=0;i<n;i++){
        cout<<"Enter Process: ";
        int x;cin>>x;
        cout<<"Enter Bust: ";
        int y;cin>>y;
        cout<<"Enter Priority: ";
        int p;cin>>p;
        arr.push_back({x,y,p,0,0});
    }
    priorityScheduling(arr);

}

File: Priority CPU Scheduling.py
class Priority:
    def __init__(self,arr):
        self.arr = arr

    def waiting(self):
        self.arr[0][3] = 0
        for i in range(1,len(self.arr)):
            self.arr[i][3] = self.arr[i-1][1] + self.arr[i-1][3]
        # return self.arr

    def turnAround(self):
        for i in range(len(self.arr)):
            self.arr[i][4] = self.arr[i][1] + self.arr[i][3]
        # return self.arr

    def priorityScheduling(self):
        # print(self.arr)
        self.arr = sorted(self.arr,key=lambda x:(x[2]),reverse=True)
        # print(self.arr)
        self.waiting()
        # print(self.arr)
        self.turnAround()
        # print(self.arr)
        tot_wt,tot_tat = 0,0
        print('Process || Bust || Priority || Waiting || Turn Around')
        for i in range(len(self.arr)):
            tot_wt += self.arr[i][3]
            tot_tat += self.arr[i][4]
            print(f'   {self.arr[i][0]}\t   {self.arr[i][1]}\t\t{self.arr[i][2]}\t  {self.arr[i][3]}\t  {self.arr[i][4]}')
        print(f'Average Waiting Time :->{tot_wt/len(self.arr)}')
        print(f'Average Turn Around Time :->{tot_tat/len(self.arr)}')

if __name__=='__main__':
    arr = []
    n = int(input('Enter Number Of Process:->'))
    for i in range(n):
        temp = []
        temp.append(int(input('Enter Process ID: ')))
        temp.append(int(input('Enter Bust : ')))
        temp.append(int(input('Enter Priority : ')))
        temp.append(0)
        temp.append(0)
        # print(temp)
        arr.append(temp.copy())
    
    # print(arr)
    obj = Priority(arr)
    obj.priorityScheduling()